Student ID: s225697234 <br>
Student Name: Ankur Luthra <br>

Part 1

Answer 1.6

In [74]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
    

In [75]:
df_meta_review_business = pd.read_csv('C:\\Users\\ANKUR\\DataScience_Material\\MDS_Deakins\\SIG742 - Modern Data Science\\SIT742Task2\\meta-review-business.csv')
df_meta_review_business.head()



name  \
0  Bear Creek Cabins & RV Park   
1             Anchorage Market   
2              Happy Camper RV   
3                 Cajun Corner   
4      Alaska General Seafoods   

                                             address  \
0  Bear Creek Cabins & RV Park, 3181 Richardson H...   
1    Anchorage Market, 88th Ave, Anchorage, AK 99515   
2  Happy Camper RV, 1151 N Shenandoah Dr # 4, Pal...   
3        Cajun Corner, 302 G St, Anchorage, AK 99501   
4  Alaska General Seafoods, 980 Stedman St, Ketch...   

                                 gmap_id description   latitude   longitude  \
0  0x56b646ed2220b77f:0xd8975e316de80952         NaN  61.100644 -146.214552   
1  0x56c8992b5dee7225:0x9f7f4bf151868cf7         NaN  61.141435 -149.868482   
2  0x56c8e0455225be87:0xf24828df75e2f8ae         NaN  61.591856 -149.290657   
3  0x56c8bdb5d91017cd:0xca19fd9afceed343         NaN  61.219378 -149.895852   
4  0x540c251956395673:0x16f5a4fe26c18931         NaN  55.336119 -131.630669   

                                           category  avg_rating  \
0  ['RV park', 'Cabin rental agency', 'Campground']         4.5   
1                               ["Farmers' market"]         4.2   
2                                ['RV repair shop']         4.4   
3                           ['American restaurant']         4.5   
4                    ['Seafood wholesaler', 'Food']         4.7   

   num_of_reviews price                                              hours  \
0              18   NaN                                                NaN   
1              18   NaN  [['Thursday', 'Closed'], ['Friday', '10AM–5PM'...   
2              28   NaN                                                NaN   
3              24   NaN  [['Wednesday', '11AM–2PM'], ['Thursday', '11AM...   
4               8   NaN  [['Wednesday', '7AM–11PM'], ['Thursday', '7AM–...   

                                                MISC                    state  \
0                                                NaN                      NaN   
1  {'Service options': ['In-store shopping'], 'Ac...  Closed ⋅ Opens 10AM Fri   
2  {'Accessibility': ['Wheelchair accessible entr...                      NaN   
3  {'Service options': ['Takeout', 'Dine-in', 'De...  Closed ⋅ Opens 11AM Thu   
4                                                NaN       Open ⋅ Closes 11PM   

                                    relative_results  \
0  ['0x56b6445fd9f9e387:0x6dd3d374ef56431a', '0x5...   
1                                                NaN   
2  ['0x56c8e104d9929a1d:0x2070ad63defadbf', '0x56...   
3                                                NaN   
4  ['0x540c25a882a72685:0xac5663d19d0a1893', '0x5...   

                                                 url  
0  https://www.google.com/maps/place//data=!4m2!3...  
1  https://www.google.com/maps/place//data=!4m2!3...  
2  https://www.google.com/maps/place//data=!4m2!3...  
3  https://www.google.com/maps/place//data=!4m2!3...  
4  https://www.google.com/maps/place//data=!4m2!3...

In [76]:
#countt the number of rows in the dataset
df_meta_review_business.shape


(12774, 15)

In [77]:
df_meta_review_business.info( )

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12774 entries, 0 to 12773
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   name              12774 non-null  object 
 1   address           12606 non-null  object 
 2   gmap_id           12774 non-null  object 
 3   description       1516 non-null   object 
 4   latitude          12774 non-null  float64
 5   longitude         12774 non-null  float64
 6   category          12712 non-null  object 
 7   avg_rating        12774 non-null  float64
 8   num_of_reviews    12774 non-null  int64  
 9   price             1585 non-null   object 
 10  hours             8684 non-null   object 
 11  MISC              9387 non-null   object 
 12  state             8062 non-null   object 
 13  relative_results  10966 non-null  object 
 14  url               12774 non-null  object 
dtypes: float64(3), int64(1), object(11)
memory usage: 1.5+ MB


### 📊 Business Meta-Information Dataset (`meta-review-business.csv`)

This dataset contains details about businesses listed on Google Maps.  
Each row represents **one business**.

| Column            | Description                                                                |
|-------------------|----------------------------------------------------------------------------|
| `name`            | Business name                                                              |
| `address`         | Street address (some missing values)                                       |
| `gmap_id`         | Unique Google Maps identifier (primary key to join with reviews)           |
| `description`     | Business description (only ~1.5k businesses have it)                       |
| `latitude`        | Latitude coordinate of the business                                        |
| `longitude`       | Longitude coordinate of the business                                       |
| `category`        | Type(s) of business (e.g., Restaurant, Cafe, RV Park)                      |
| `avg_rating`      | Average rating of the business (float)                                     |
| `num_of_reviews`  | Total number of reviews submitted                                          |
| `price`           | Price range indicator (e.g., $, $$, $$$) — very sparse (~1.5k rows)        |
| `hours`           | Business working hours (~8.6k rows filled)                                 |
| `MISC`            | Miscellaneous info (e.g., service options, accessibility)                  |
| `state`           | State/region or current status (e.g., "Closed · Opens 10AM Fri")           |
| `relative_results`| Related businesses suggested by data provider                              |
| `url`             | Direct Google Maps business link                                           |

**Shape:** 12,774 rows × 15 columns  


In [78]:
df_review = pd.read_csv('C:\\Users\\ANKUR\\DataScience_Material\\MDS_Deakins\\SIG742 - Modern Data Science\\SIT742Task2\\review.csv')

df_review.head()





user_id               name           time  rating  \
0  1.091298e+20         Nicki Gore  1566331951619       5   
1  1.132409e+20      Allen Ratliff  1504917982385       5   
2  1.130448e+20  Jonathan Tringali  1474765901185       4   
3  1.103292e+20             S Blad  1472858535682       4   
4  1.089896e+20     Daniel Formoso  1529649811341       5   

                                                text pics resp  \
0  We always stay here when in Valdez for silver ...  NaN  NaN   
1  Great campground for the price. Nice hot unlim...  NaN  NaN   
2  We tent camped here for 2 nights while explori...  NaN  NaN   
3  This place is just a few miles outside Valdez,...  NaN  NaN   
4  Probably the nicest and cleanest campground we...  NaN  NaN   

                                 gmap_id  
0  0x56b646ed2220b77f:0xd8975e316de80952  
1  0x56b646ed2220b77f:0xd8975e316de80952  
2  0x56b646ed2220b77f:0xd8975e316de80952  
3  0x56b646ed2220b77f:0xd8975e316de80952  
4  0x56b646ed2220b77f:0xd8975e316de80952

In [79]:
#find number of unique gmap_id in the review dataset
df_review['gmap_id'].nunique()


6019

In [80]:

#find number of reviews per gmap_id
df_review['gmap_id'].value_counts()


gmap_id
0x56c897b9ce6000df:0xd707e127588a8c6c    2833
0x56c899d058920487:0x12882cc9133f2f54    2594
0x56c897c63697ee39:0x419904ababbc740b    2258
0x56c8965ee2fb87a1:0x559736347bd48842    2237
0x56c89629bde7481f:0x7e8a9413ab25d5d     2219
                                         ... 
0x56c897c82598d177:0xe8cb4060773c28bf      10
0x56c8dfd605198db1:0x23e90e52d496197a      10
0x56c8de1a70f260b9:0x2e9674ca6dae963c      10
0x56c9299752ca9635:0xa2ad5d2ad71a3f1a      10
0x56c864cabe37de37:0x15efe7dcb7ce0276      10
Name: count, Length: 6019, dtype: int64

In [81]:
df_review.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 521515 entries, 0 to 521514
Data columns (total 8 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   user_id  521515 non-null  float64
 1   name     521515 non-null  object 
 2   time     521515 non-null  int64  
 3   rating   521515 non-null  int64  
 4   text     298250 non-null  object 
 5   pics     20851 non-null   object 
 6   resp     44232 non-null   object 
 7   gmap_id  521515 non-null  object 
dtypes: float64(1), int64(2), object(5)
memory usage: 31.8+ MB


### 📝 Review Dataset (`review.csv`)

This dataset contains user-submitted reviews for businesses on Google Maps.  
Each row represents **one review**.

| Column      | Description                                                                 |
|-------------|-----------------------------------------------------------------------------|
| `user_id`   | Unique identifier for the reviewer (float, anonymized account ID)    |
| `name`      | Name of the reviewer                                                        |
| `time`      | Unix timestamp of when the review was posted (convertible to datetime)      |
| `rating`    | User’s star rating for the business (integer: 1–5)                          |
| `text`      | Written review content (only ~298k reviews have text)                       |
| `pics`      | Links/paths to uploaded pictures (~20k reviews contain pictures)            |
| `resp`      | Business owner’s response to the review (~44k reviews contain responses)    |
| `gmap_id`   | Business identifier (foreign key — links to `meta-review-business.csv`)     |

**Shape:** 521,515 rows × 8 columns  


In [82]:
df_review["time"] = pd.to_datetime(df_review["time"], unit="ms")

In [83]:
#find all reviews for gmap_id 0x56c864cabe37de37:0x15efe7dcb7ce0276
df_review[df_review['gmap_id'] == '0x56c864cabe37de37:0x15efe7dcb7ce0276']

user_id                        name                    time  \
94807  1.118296e+20         Marie Josee Lamothe 2019-08-01 16:14:05.523   
94808  1.085611e+20                  Amie Kirby 2020-08-16 17:48:15.996   
94809  1.027337e+20                  Gary Poppe 2019-05-13 23:26:25.434   
94810  1.018784e+20             Sandra Garrison 2018-08-05 21:36:20.350   
94811  1.039698e+20                   Brian Won 2018-06-13 01:23:39.988   
94812  1.118194e+20               Taylor Kelley 2016-05-03 21:42:09.323   
94813  1.029747e+20                   Richard M 2016-05-18 22:23:41.148   
94814  1.177482e+20  Bacon & Cheese Productions 2016-07-01 07:26:35.151   
94815  1.168244e+20                  Ty Bridges 2017-04-30 01:46:50.973   
94816  1.057113e+20               Walden Walden 2018-05-23 05:50:44.722   

       rating                                               text  \
94807       4  Very professional and friendly staff. Friendly...   
94808       5  Helicopter ride was excellent. The captain was...   
94809       5  Amazing experience and awesome staff.  Well wo...   
94810       5  This was our first excursion in AK and maybe o...   
94811       5  Spotless helicopters, great staff, and excelle...   
94812       5  Amazing flight seeing tours daily. Took my par...   
94813       5  Did the dog tour and it was amazing. Flying up...   
94814       5  Best thing to do in Girdwood, if you can affor...   
94815       5  Great pilots and staff! Friendly, knowledgeabl...   
94816       5  (Translated by Google) It has really been a un...   

                                                    pics  \
94807  [{'url': ['https://lh5.googleusercontent.com/p...   
94808                                                NaN   
94809                                                NaN   
94810                                                NaN   
94811                                                NaN   
94812                                                NaN   
94813                                                NaN   
94814                                                NaN   
94815                                                NaN   
94816                                                NaN   

                                                    resp  \
94807                                                NaN   
94808  {'time': 1597685390414, 'text': 'Amie, thank y...   
94809                                                NaN   
94810                                                NaN   
94811                                                NaN   
94812                                                NaN   
94813                                                NaN   
94814                                                NaN   
94815                                                NaN   
94816                                                NaN   

                                     gmap_id  
94807  0x56c864cabe37de37:0x15efe7dcb7ce0276  
94808  0x56c864cabe37de37:0x15efe7dcb7ce0276  
94809  0x56c864cabe37de37:0x15efe7dcb7ce0276  
94810  0x56c864cabe37de37:0x15efe7dcb7ce0276  
94811  0x56c864cabe37de37:0x15efe7dcb7ce0276  
94812  0x56c864cabe37de37:0x15efe7dcb7ce0276  
94813  0x56c864cabe37de37:0x15efe7dcb7ce0276  
94814  0x56c864cabe37de37:0x15efe7dcb7ce0276  
94815  0x56c864cabe37de37:0x15efe7dcb7ce0276  
94816  0x56c864cabe37de37:0x15efe7dcb7ce0276

In [84]:
df_review.shape

(521515, 8)

In [85]:
review_business_meta_merged_df = df_review.merge(df_meta_review_business, on="gmap_id", how="left")


In [86]:
review_business_meta_merged_df.head()

user_id             name_x                    time  rating  \
0  1.091298e+20         Nicki Gore 2019-08-20 20:12:31.619       5   
1  1.091298e+20         Nicki Gore 2019-08-20 20:12:31.619       5   
2  1.132409e+20      Allen Ratliff 2017-09-09 00:46:22.385       5   
3  1.132409e+20      Allen Ratliff 2017-09-09 00:46:22.385       5   
4  1.130448e+20  Jonathan Tringali 2016-09-25 01:11:41.185       4   

                                                text pics resp  \
0  We always stay here when in Valdez for silver ...  NaN  NaN   
1  We always stay here when in Valdez for silver ...  NaN  NaN   
2  Great campground for the price. Nice hot unlim...  NaN  NaN   
3  Great campground for the price. Nice hot unlim...  NaN  NaN   
4  We tent camped here for 2 nights while explori...  NaN  NaN   

                                 gmap_id                       name_y  \
0  0x56b646ed2220b77f:0xd8975e316de80952  Bear Creek Cabins & RV Park   
1  0x56b646ed2220b77f:0xd8975e316de80952  Bear Creek Cabins & RV Park   
2  0x56b646ed2220b77f:0xd8975e316de80952  Bear Creek Cabins & RV Park   
3  0x56b646ed2220b77f:0xd8975e316de80952  Bear Creek Cabins & RV Park   
4  0x56b646ed2220b77f:0xd8975e316de80952  Bear Creek Cabins & RV Park   

                                             address  ...   longitude  \
0  Bear Creek Cabins & RV Park, 3181 Richardson H...  ... -146.214552   
1  Bear Creek Cabins & RV Park, 3181 Richardson H...  ... -146.214552   
2  Bear Creek Cabins & RV Park, 3181 Richardson H...  ... -146.214552   
3  Bear Creek Cabins & RV Park, 3181 Richardson H...  ... -146.214552   
4  Bear Creek Cabins & RV Park, 3181 Richardson H...  ... -146.214552   

                                           category  avg_rating  \
0  ['RV park', 'Cabin rental agency', 'Campground']         4.5   
1  ['RV park', 'Cabin rental agency', 'Campground']         4.5   
2  ['RV park', 'Cabin rental agency', 'Campground']         4.5   
3  ['RV park', 'Cabin rental agency', 'Campground']         4.5   
4  ['RV park', 'Cabin rental agency', 'Campground']         4.5   

  num_of_reviews  price  hours MISC state  \
0             18    NaN    NaN  NaN   NaN   
1             18    NaN    NaN  NaN   NaN   
2             18    NaN    NaN  NaN   NaN   
3             18    NaN    NaN  NaN   NaN   
4             18    NaN    NaN  NaN   NaN   

                                    relative_results  \
0  ['0x56b6445fd9f9e387:0x6dd3d374ef56431a', '0x5...   
1  ['0x56b6445fd9f9e387:0x6dd3d374ef56431a', '0x5...   
2  ['0x56b6445fd9f9e387:0x6dd3d374ef56431a', '0x5...   
3  ['0x56b6445fd9f9e387:0x6dd3d374ef56431a', '0x5...   
4  ['0x56b6445fd9f9e387:0x6dd3d374ef56431a', '0x5...   

                                                 url  
0  https://www.google.com/maps/place//data=!4m2!3...  
1  https://www.google.com/maps/place//data=!4m2!3...  
2  https://www.google.com/maps/place//data=!4m2!3...  
3  https://www.google.com/maps/place//data=!4m2!3...  
4  https://www.google.com/maps/place//data=!4m2!3...  

[5 rows x 22 columns]

In [87]:
# find any rows in df_meta_review_business that are not in df_review
df_meta_review_business[~df_meta_review_business['gmap_id'].isin(df_review['gmap_id'])]

name  \
3                            Cajun Corner   
4                 Alaska General Seafoods   
6                     Silver Bay Seafoods   
7      A Room With A View Bed & Breakfast   
10               Coastal Cold Storage Inc   
...                                   ...   
12725                  Rising Tide Bakery   
12727                        Grace Harbor   
12729       Old Sitka State Historic Park   
12734      Halibut Point Crab and Chowder   
12752       Glacier City Snowmobile Tours   

                                                 address  \
3            Cajun Corner, 302 G St, Anchorage, AK 99501   
4      Alaska General Seafoods, 980 Stedman St, Ketch...   
6      Silver Bay Seafoods, 622 Frontier St, Naknek, ...   
7      A Room With A View Bed & Breakfast, 840 Rosebu...   
10     Coastal Cold Storage Inc, 306 N Nordic Dr, Pet...   
...                                                  ...   
12725  Rising Tide Bakery, 143 Price St bldg a, Sitka...   
12727  Grace Harbor, 1904 Halibut Point Rd, Sitka, AK...   
12729  Old Sitka State Historic Park, Old Sitka Histo...   
12734  Halibut Point Crab and Chowder, 4513 Halibut P...   
12752  Glacier City Snowmobile Tours, 36511 Seward Hw...   

                                     gmap_id description   latitude  \
3      0x56c8bdb5d91017cd:0xca19fd9afceed343         NaN  61.219378   
4      0x540c251956395673:0x16f5a4fe26c18931         NaN  55.336119   
6      0x56e7db2908155f81:0xc57ff97f25ffe1ce         NaN  58.743921   
7      0x56c132808b88408d:0x138bcd1709abfdea         NaN  59.665086   
10     0x5406038d706891a9:0x3b0075d07a52d830         NaN  56.813490   
...                                      ...         ...        ...   
12725  0x54046722afcd2cc3:0xc8dd52c9d1a580ea         NaN  57.048777   
12727  0x54046609e68fb95b:0x30b9d701ff6c4cb5         NaN  57.069629   
12729  0x54046f45c710e2a7:0x4c52a8a3a09a23b5         NaN  57.129435   
12734  0x540465f845cb7eff:0x3f11aaa9790a121b         NaN  57.115478   
12752  0x56c87083e907f055:0xdcf113407e45ba22         NaN  60.940796   

        longitude                              category  avg_rating  \
3     -149.895852               ['American restaurant']         4.5   
4     -131.630669        ['Seafood wholesaler', 'Food']         4.7   
6     -156.953535                   ['Fish processing']         4.1   
7     -151.520351                   ['Bed & breakfast']         5.0   
10    -132.957227                ['Seafood wholesaler']         4.8   
...           ...                                   ...         ...   
12725 -135.305528                            ['Bakery']         5.0   
12727 -135.366079                    ['Baptist church']         4.9   
12729 -135.373651  ['State park', 'Tourist attraction']         4.8   
12734 -135.393167  ['Seafood restaurant', 'Restaurant']         4.6   
12752 -149.171646                       ['Tour agency']         4.7   

       num_of_reviews price  \
3                  24   NaN   
4                   8   NaN   
6                  18   NaN   
7                   3   NaN   
10                 18   NaN   
...               ...   ...   
12725               5   NaN   
12727               8   NaN   
12729              18   NaN   
12734              26   NaN   
12752              48   NaN   

                                                   hours  \
3      [['Wednesday', '11AM–2PM'], ['Thursday', '11AM...   
4      [['Wednesday', '7AM–11PM'], ['Thursday', '7AM–...   
6                                                    NaN   
7                                                    NaN   
10     [['Wednesday', '6:30AM–6PM'], ['Thursday', '6:...   
...                                                  ...   
12725  [['Saturday', 'Closed'], ['Sunday', '9AM–12PM'...   
12727                                                NaN   
12729  [['Saturday', 'Open 24 hours'], ['Sunday', 'Op...   
12734                                                NaN   
12752  [['Thursday', '7AM–8PM'], ['F

In [88]:
business_meta_review_merged_df = df_meta_review_business.merge(df_review, on="gmap_id", how="left")

In [89]:
#find all rows where user_id is null
business_meta_review_merged_df[business_meta_review_merged_df['user_id'].isnull()]

name_x  \
46                            Cajun Corner   
47                 Alaska General Seafoods   
58                     Silver Bay Seafoods   
59      A Room With A View Bed & Breakfast   
100               Coastal Cold Storage Inc   
...                                    ...   
513895                  Rising Tide Bakery   
513915                        Grace Harbor   
513975       Old Sitka State Historic Park   
514088      Halibut Point Crab and Chowder   
522099       Glacier City Snowmobile Tours   

                                                  address  \
46            Cajun Corner, 302 G St, Anchorage, AK 99501   
47      Alaska General Seafoods, 980 Stedman St, Ketch...   
58      Silver Bay Seafoods, 622 Frontier St, Naknek, ...   
59      A Room With A View Bed & Breakfast, 840 Rosebu...   
100     Coastal Cold Storage Inc, 306 N Nordic Dr, Pet...   
...                                                   ...   
513895  Rising Tide Bakery, 143 Price St bldg a, Sitka...   
513915  Grace Harbor, 1904 Halibut Point Rd, Sitka, AK...   
513975  Old Sitka State Historic Park, Old Sitka Histo...   
514088  Halibut Point Crab and Chowder, 4513 Halibut P...   
522099  Glacier City Snowmobile Tours, 36511 Seward Hw...   

                                      gmap_id description   latitude  \
46      0x56c8bdb5d91017cd:0xca19fd9afceed343         NaN  61.219378   
47      0x540c251956395673:0x16f5a4fe26c18931         NaN  55.336119   
58      0x56e7db2908155f81:0xc57ff97f25ffe1ce         NaN  58.743921   
59      0x56c132808b88408d:0x138bcd1709abfdea         NaN  59.665086   
100     0x5406038d706891a9:0x3b0075d07a52d830         NaN  56.813490   
...                                       ...         ...        ...   
513895  0x54046722afcd2cc3:0xc8dd52c9d1a580ea         NaN  57.048777   
513915  0x54046609e68fb95b:0x30b9d701ff6c4cb5         NaN  57.069629   
513975  0x54046f45c710e2a7:0x4c52a8a3a09a23b5         NaN  57.129435   
514088  0x540465f845cb7eff:0x3f11aaa9790a121b         NaN  57.115478   
522099  0x56c87083e907f055:0xdcf113407e45ba22         NaN  60.940796   

         longitude                              category  avg_rating  \
46     -149.895852               ['American restaurant']         4.5   
47     -131.630669        ['Seafood wholesaler', 'Food']         4.7   
58     -156.953535                   ['Fish processing']         4.1   
59     -151.520351                   ['Bed & breakfast']         5.0   
100    -132.957227                ['Seafood wholesaler']         4.8   
...            ...                                   ...         ...   
513895 -135.305528                            ['Bakery']         5.0   
513915 -135.366079                    ['Baptist church']         4.9   
513975 -135.373651  ['State park', 'Tourist attraction']         4.8   
514088 -135.393167  ['Seafood restaurant', 'Restaurant']         4.6   
522099 -149.171646                       ['Tour agency']         4.7   

        num_of_reviews price  ...                    state  \
46                  24   NaN  ...  Closed ⋅ Opens 11AM Thu   
47                   8   NaN  ...       Open ⋅ Closes 11PM   
58                  18   NaN  ...                      NaN   
59                   3   NaN  ...                      NaN   
100                 18   NaN  ...        Open ⋅ Closes 6PM   
...                ...   ...  ...                      ...   
513895               5   NaN  ...                      NaN   
513915               8   NaN  ...                      NaN   
513975              18   NaN  ...                      NaN   
514088              26   NaN  ...                      NaN   
522099              48   NaN  ...                      NaN   

                                         relative_results  \
46                                                    NaN   
47      ['0x540c25a882a72685:0xac5663d19d0a1893', '0x5...   
58      ['0x56e7c547466b5c23:0x5bae1257a2441fd9', '0x5...   
59                             

In [ ]:


import ast
import numpy as np
import pandas as pd
from scipy import sparse
from sklearn.preprocessing import MultiLabelBinarizer, normalize
from sklearn.neighbors import NearestNeighbors

# ---------- 1) Load & preprocess data ----------
def load_and_preprocess(business_csv="meta-review-business.csv", review_csv="review.csv"):
    # Load
    business_df = pd.read_csv(business_csv)
    review_df = pd.read_csv(review_csv)
    
    # Convert review "time" (likely ms since epoch) to datetime
    try:
        review_df['time'] = pd.to_datetime(review_df['time'], unit='ms')
    except Exception:
        # fallback (if already seconds)
        review_df['time'] = pd.to_datetime(review_df['time'], unit='s', errors='coerce')
    
    # Normalize gmap_id to string
    business_df['gmap_id'] = business_df['gmap_id'].astype(str)
    review_df['gmap_id'] = review_df['gmap_id'].astype(str)
    
    # Parse categories (category may be stored as a string-repr list)
    def safe_parse_list(x):
        if pd.isna(x):
            return []
        if isinstance(x, list):
            return x
        try:
            # Some rows use bracket-list strings, others may be simple text
            parsed = ast.literal_eval(x)
            # if parsed is a string, wrap it
            if isinstance(parsed, str):
                return [parsed]
            return parsed
        except Exception:
            # fallback: split on comma if looks like "A, B"
            if isinstance(x, str):
                if ',' in x:
                    return [s.strip().strip("'\"") for s in x.split(',') if s.strip()]
                else:
                    return [x.strip()]
            return []
    
    business_df['categories_parsed'] = business_df['category'].apply(safe_parse_list)
    
    # Aggregate reviews: if there are duplicates (same user_id, same gmap_id),
    # compute the mean rating or take the latest rating. We'll use mean here.
    # Also drop reviews with missing rating
    reviews = review_df.dropna(subset=['rating']).copy()
    # Ensure user_id is string (some are floats/scientific)
    reviews['user_id'] = reviews['user_id'].astype(str)
    # Aggregate
    reviews_agg = reviews.groupby(['user_id', 'gmap_id'], as_index=False).agg({
        'rating': 'mean',  # if multiple, average
        'time': 'max'      # keep the latest time (optional)
    })
    
    return business_df, reviews_agg

# ---------- 2) Build item-user matrix ----------
def build_item_user_matrix(reviews, business_df,
                           min_user_ratings=5, min_item_ratings=10,
                           keep_businesses_without_reviews=False):
    """
    reviews: DataFrame with columns ['user_id','gmap_id','rating']
    business_df: DataFrame with 'gmap_id'
    returns:
      - item_user_matrix (scipy.sparse.csr_matrix) shape = (n_items, n_users)
      - item_index_to_gmap (list), gmap_to_item_index (dict)
      - user_index_to_id (list), user_id_to_index (dict)
      - filtered_business_df (DataFrame) aligned to item indices
    """
    # Count user and item activity
    user_counts = reviews['user_id'].value_counts()
    item_counts = reviews['gmap_id'].value_counts()
    
    # Keep only users and items with enough activity
    keep_users = set(user_counts[user_counts >= min_user_ratings].index)
    keep_items = set(item_counts[item_counts >= min_item_ratings].index)
    
    reviews_f = reviews[reviews['user_id'].isin(keep_users) & reviews['gmap_id'].isin(keep_items)].copy()
    
    # Build index mapping for users and items (only those kept)
    user_ids = sorted(reviews_f['user_id'].unique())
    item_ids = sorted(reviews_f['gmap_id'].unique())
    user_id_to_index = {uid:i for i,uid in enumerate(user_ids)}
    item_id_to_index = {iid:i for i,iid in enumerate(item_ids)}
    
    # Build COO arrays
    rows = reviews_f['gmap_id'].map(item_id_to_index).to_numpy()
    cols = reviews_f['user_id'].map(user_id_to_index).to_numpy()
    data = reviews_f['rating'].to_numpy(dtype=float)
    
    # Create sparse matrix (items x users)
    item_user_coo = sparse.coo_matrix((data, (rows, cols)), shape=(len(item_ids), len(user_ids)))
    item_user = item_user_coo.tocsr()
    
    # Optionally include businesses that were filtered out (no/low reviews) into business_df
    filtered_business_df = business_df[business_df['gmap_id'].isin(item_ids)].copy()
    filtered_business_df = filtered_business_df.set_index('gmap_id').loc[item_ids].reset_index()
    
    return item_user, item_ids, user_ids, item_id_to_index, user_id_to_index, filtered_business_df

# ---------- 3) Build KNN graph (top-K neighbors per item) ----------
def build_item_knn_graph(item_user_matrix, n_neighbors=50, metric='cosine', n_jobs=1):
    """
    Build top-K nearest neighbors for each item using item vectors (rows of item_user_matrix).
    Returns a sparse similarity matrix S (CSR) where S[i,j] = similarity between item i and j (nonzero only for top-K neighbors).
    """
    # Use NearestNeighbors with metric='cosine' (returns distances in [0,2], where 0 is identical)
    nn = NearestNeighbors(n_neighbors=n_neighbors+1, metric=metric, algorithm='brute', n_jobs=n_jobs)
    nn.fit(item_user_matrix)  # items are rows
    
    distances, neighbors = nn.kneighbors(item_user_matrix, return_distance=True)
    # distances shape = (n_items, n_neighbors+1) - first neighbor is the item itself (distance 0)
    n_items = item_user_matrix.shape[0]
    
    rows = []
    cols = []
    data = []
    
    for i in range(n_items):
        for dist, nbr in zip(distances[i,1:], neighbors[i,1:]):  # skip self (index 0)
            sim = 1.0 - dist  # cosine similarity approximate (since dist = 1 - cosine) but sklearn returns distance=cosine
            if sim > 0:
                rows.append(i)
                cols.append(nbr)
                data.append(sim)
    # Build sparse matrix item x item
    S = sparse.csr_matrix((data, (rows, cols)), shape=(n_items, n_items))
    # Optionally normalize rows (so predictions are weighted averages)
    S = normalize(S, norm='l1', axis=1)  # row-stochastic: row sums = 1
    return S

# ---------- 4) Content (category) similarity ----------
def build_category_similarity(business_df_filtered, top_k=50):
    """
    business_df_filtered must be aligned to item_ids (order matters).
    Returns a row-normalized sparse similarity matrix C (items x items) built from categories.
    """
    mlb = MultiLabelBinarizer(sparse_output=True)
    cat_matrix = mlb.fit_transform(business_df_filtered['categories_parsed'].apply(lambda x: x if isinstance(x, (list,tuple)) else []))
    # cat_matrix is sparse (items x n_categories)
    # Compute cosine similarity via NearestNeighbors to get top_k neighbors for each item
    nn = NearestNeighbors(n_neighbors=min(top_k+1, cat_matrix.shape[0]), metric='cosine', algorithm='brute')
    nn.fit(cat_matrix)
    distances, neighbors = nn.kneighbors(cat_matrix, return_distance=True)
    
    rows, cols, data = [], [], []
    for i in range(cat_matrix.shape[0]):
        for dist, nbr in zip(distances[i,1:], neighbors[i,1:]):
            sim = 1.0 - dist
            if sim > 0:
                rows.append(i); cols.append(nbr); data.append(sim)
    C = sparse.csr_matrix((data, (rows, cols)), shape=(cat_matrix.shape[0], cat_matrix.shape[0]))
    C = normalize(C, norm='l1', axis=1)
    return C

# ---------- 5) Combine similarity graphs (hybrid) ----------
def combine_similarity_graphs(S_collab, S_content, alpha=0.7):
    """
    alpha: weight for collaborative similarity. (1-alpha) weight for content similarity.
    S_* are row-normalized sparse matrices (CSR).
    Returns row-normalized combined S.
    """
    # Convert to same sparse format and combine
    S = alpha * S_collab + (1.0 - alpha) * S_content
    S = normalize(S, norm='l1', axis=1)
    return S

# ---------- 6) Recommendation functions ----------
def recommend_similar_businesses(gmap_id, item_ids, business_df_filtered, S_combined,
                                 top_n=10, exclude_self=True):
    """
    Recommend businesses similar to the given gmap_id using the combined similarity graph.
    """
    if gmap_id not in item_ids:
        return pd.DataFrame()  # fallback: empty (business not in collab space)
    idx = item_ids.index(gmap_id)
    row = S_combined.getrow(idx).toarray().ravel()
    # rank
    top_idx = np.argsort(row)[::-1]
    if exclude_self:
        top_idx = top_idx[top_idx != idx]
    top_idx = top_idx[:top_n]
    recs = business_df_filtered.iloc[top_idx].copy()
    recs['similarity_score'] = row[top_idx]
    return recs[['gmap_id','name','address','categories_parsed','avg_rating','num_of_reviews','similarity_score']]

def predict_for_user(user_id, user_id_to_index, item_ids, S_combined, item_user_matrix, top_n=10):
    """
    Predict scores for a user using item-based CF:
      pred_item_scores = S_combined dot user_ratings_vector
    Returns top_n recommendations (gmap_id, predicted_score).
    """
    if user_id not in user_id_to_index:
        return pd.DataFrame()  # cold-start user fallback outside this filtered set
    uidx = user_id_to_index[user_id]
    # user ratings vector aligned to columns of item_user_matrix (items x users)
    user_ratings = item_user_matrix[:, uidx].toarray().ravel()  # shape n_items
    # compute predicted scores
    pred_scores = S_combined.dot(user_ratings)
    # Do not recommend items the user already rated
    already_rated = user_ratings > 0
    pred_scores[already_rated] = -np.inf
    top_idx = np.argsort(pred_scores)[::-1][:top_n]
    res = pd.DataFrame({
        'gmap_id': [item_ids[i] for i in top_idx],
        'predicted_score': pred_scores[top_idx]
    })
    return res

# ---------- 7) Putting it all together ----------
def build_and_run_pipeline(business_csv="meta-review-business.csv", review_csv="review.csv",
                           min_user_ratings=5, min_item_ratings=10,
                           knn_k=50, cat_k=50, hybrid_alpha=0.7, n_jobs=1):
    business_df, reviews_agg = load_and_preprocess(business_csv, review_csv)
    print("Loaded: businesses =", len(business_df), "reviews (agg rows) =", len(reviews_agg))
    item_user, item_ids, user_ids, item_id_to_index, user_id_to_index, business_filtered = build_item_user_matrix(
        reviews_agg, business_df, min_user_ratings=min_user_ratings, min_item_ratings=min_item_ratings
    )
    print("Filtered item_user matrix:", item_user.shape, "items kept:", len(item_ids), "users kept:", len(user_ids))
    
    # Build collaborative KNN graph
    print("Building collaborative KNN graph (k=%d)..." % knn_k)
    S_collab = build_item_knn_graph(item_user, n_neighbors=knn_k, n_jobs=n_jobs)
    
    # Build category similarity
    print("Building category similarity (k=%d)..." % cat_k)
    S_cat = build_category_similarity(business_filtered, top_k=cat_k)
    
    # Combine
    # print("Combining graphs alpha=", hybrid_alpha)
    # S_combined = combine_similarity_graphs(S_collab, S_cat, alpha=hybrid_alpha)
    
    # return {
    #     'business_df': business_df,
    #     'reviews_agg': reviews_agg,
    #     'item_user_matrix': item_user,
    #     'item_ids': item_ids,
    #     'user_ids': user_ids,
    #     'item_id_to_index': item_id_to_index,
    #     'user_id_to_index': user_id_to_index,
    #     'business_filtered': business_filtered,
    #     'S_collab': S_collab,
    #     'S_cat': S_cat,
    #     'S_combined': S_combined
    # }




In [105]:
business_csv="meta-review-business.csv"
review_csv="review.csv"
business_df, reviews_agg = load_and_preprocess(business_csv, review_csv)

In [106]:
reviews_agg.head()

user_id                                gmap_id  rating  \
0  1.0000266958784964e+20   0x56c89629bde7481f:0x7e8a9413ab25d5d     4.0   
1  1.0000266958784964e+20  0x56c8965d14c17b89:0x5ded6a79373127da     5.0   
2  1.0000266958784964e+20  0x56c8965ea5cfc777:0xf4e6f66b783ae126     4.0   
3  1.0000266958784964e+20  0x56c8965ee2fb87a1:0x559736347bd48842     5.0   
4  1.0000266958784964e+20  0x56c89799e5eb2a5f:0xcd38ccf0138aa3b5     4.0   

                     time  
0 2019-04-29 02:13:43.775  
1 2019-03-11 03:28:24.124  
2 2019-11-20 03:01:23.562  
3 2019-11-20 03:02:00.638  
4 2019-07-19 18:59:26.225

In [96]:
business_df.head()

name  \
0  Bear Creek Cabins & RV Park   
1             Anchorage Market   
2              Happy Camper RV   
3                 Cajun Corner   
4      Alaska General Seafoods   

                                             address  \
0  Bear Creek Cabins & RV Park, 3181 Richardson H...   
1    Anchorage Market, 88th Ave, Anchorage, AK 99515   
2  Happy Camper RV, 1151 N Shenandoah Dr # 4, Pal...   
3        Cajun Corner, 302 G St, Anchorage, AK 99501   
4  Alaska General Seafoods, 980 Stedman St, Ketch...   

                                 gmap_id description   latitude   longitude  \
0  0x56b646ed2220b77f:0xd8975e316de80952         NaN  61.100644 -146.214552   
1  0x56c8992b5dee7225:0x9f7f4bf151868cf7         NaN  61.141435 -149.868482   
2  0x56c8e0455225be87:0xf24828df75e2f8ae         NaN  61.591856 -149.290657   
3  0x56c8bdb5d91017cd:0xca19fd9afceed343         NaN  61.219378 -149.895852   
4  0x540c251956395673:0x16f5a4fe26c18931         NaN  55.336119 -131.630669   

                                           category  avg_rating  \
0  ['RV park', 'Cabin rental agency', 'Campground']         4.5   
1                               ["Farmers' market"]         4.2   
2                                ['RV repair shop']         4.4   
3                           ['American restaurant']         4.5   
4                    ['Seafood wholesaler', 'Food']         4.7   

   num_of_reviews price                                              hours  \
0              18   NaN                                                NaN   
1              18   NaN  [['Thursday', 'Closed'], ['Friday', '10AM–5PM'...   
2              28   NaN                                                NaN   
3              24   NaN  [['Wednesday', '11AM–2PM'], ['Thursday', '11AM...   
4               8   NaN  [['Wednesday', '7AM–11PM'], ['Thursday', '7AM–...   

                                                MISC                    state  \
0                                                NaN                      NaN   
1  {'Service options': ['In-store shopping'], 'Ac...  Closed ⋅ Opens 10AM Fri   
2  {'Accessibility': ['Wheelchair accessible entr...                      NaN   
3  {'Service options': ['Takeout', 'Dine-in', 'De...  Closed ⋅ Opens 11AM Thu   
4                                                NaN       Open ⋅ Closes 11PM   

                                    relative_results  \
0  ['0x56b6445fd9f9e387:0x6dd3d374ef56431a', '0x5...   
1                                                NaN   
2  ['0x56c8e104d9929a1d:0x2070ad63defadbf', '0x56...   
3                                                NaN   
4  ['0x540c25a882a72685:0xac5663d19d0a1893', '0x5...   

                                                 url  \
0  https://www.google.com/maps/place//data=!4m2!3...   
1  https://www.google.com/maps/place//data=!4m2!3...   
2  https://www.google.com/maps/place//data=!4m2!3...   
3  https://www.google.com/maps/place//data=!4m2!3...   
4  https://www.google.com/maps/place//data=!4m2!3...   

                            categories_parsed  
0  [RV park, Cabin rental agency, Campground]  
1                           [Farmers' market]  
2                            [RV repair shop]  
3                       [American restaurant]  
4                  [Seafood wholesaler, Food]

In [93]:
#---------- Example usage (uncomment + run in your environment) ----------
pipeline = build_and_run_pipeline(
    business_csv="meta-review-business.csv",
    review_csv="review.csv",
    min_user_ratings=5,
    min_item_ratings=10,
    knn_k=50,
    cat_k=50,
    hybrid_alpha=0.7,
    n_jobs=4
)

# # Example 1: Recommend for a business (e.g., Bear Creek gmap id shown in your sample)
# gmap_example = "0x56b646ed2220b77f:0xd8975e316de80952"
# recs = recommend_similar_businesses(gmap_example,
#                                     pipeline['item_ids'],
#                                     pipeline['business_filtered'],
#                                     pipeline['S_combined'],
#                                     top_n=10)
# print("Recommended businesses similar to", gmap_example)
# print(recs)
#
# # Example 2: Recommend for a user (user_id must be in the filtered set)
# some_user = pipeline['user_ids'][0]  # pick one
# user_recs = predict_for_user(some_user, pipeline['user_id_to_index'],
#                              pipeline['item_ids'],
#                              pipeline['S_combined'],
#                              pipeline['item_user_matrix'], top_n=10)
# print("Recommendations for user", some_user)
# print(user_recs)
#
# # Save recommendations to CSV (optional)
# recs.to_csv("example_business_recommendations.csv", index=False)


Loaded: businesses = 12774 reviews (agg rows) = 516289
Filtered item_user matrix: (6000, 20022) items kept: 6000 users kept: 20022
Building collaborative KNN graph (k=50)...
Building category similarity (k=50)...
